core/testcontainers/core/waiting_utils.py

```diff
@@ -11,122 +11,370 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 
-
 import re
 import time
-import traceback
-from typing import TYPE_CHECKING, Any, Callable, Union
+import warnings
+from abc import ABC, abstractmethod
+from datetime import timedelta
+from typing import Any, Callable, Protocol, TypeVar, Union, cast
 
 import wrapt
 
 from testcontainers.core.config import testcontainers_config as config
 from testcontainers.core.utils import setup_logger
 
-if TYPE_CHECKING:
-    from testcontainers.core.container import DockerContainer
-
 logger = setup_logger(__name__)
 
 # Get a tuple of transient exceptions for which we'll retry. Other exceptions will be raised.
 TRANSIENT_EXCEPTIONS = (TimeoutError, ConnectionError)
 
+# Type variables for generic functions
+F = TypeVar("F", bound=Callable[..., Any])
+
 
-def wait_container_is_ready(*transient_exceptions) -> Callable:
+class WaitStrategyTarget(Protocol):
```

Comment:
I wonder if `WaitStrategyTarget` & `WaitStrategy` might be split into their own module - somewhat separating it from the now deprecated functions in here.  That might help simplify when time comes for its eventual removal.